### PR TITLE
Handle out of order workflow replication

### PR DESCRIPTION
**Pull request by wxing1292**
* Attach unique ID generated by shard to each workflow execution during creation
* All replication message will include the above ID
* When workflow with the same ID are replicated out of order, comparison of the above ID will drop the replication message with smaller ID

### CODE REVIEW

File: service/frontend/workflowHandler.go [hunk 1]
```diff
@@ -87,6 +87,7 @@ type (
 		PersistenceToken  []byte
 		TransientDecision *gen.TransientDecisionInfo
 		EventStoreVersion int32
+		CreateTaskID      int64
```
Comment (samarabbas) [inline service/frontend/workflowHandler.go]: Is this backward compatible?

Comment (wxing1292) [reply service/frontend/workflowHandler.go]: public API does not use this field,
only re-send stack use it

File: service/history/historyReplicator.go [hunk 2]
```diff
@@ -983,22 +997,28 @@ func (r *historyReplicator) replicateWorkflowStarted(ctx context.Context, contex
...
 		logError(logger, "Fail to flush buffer for current workflow.", err)
-		return "", 0, false, err
+		return "", 0, 0, false, err
 	}
-	return currentContext.getExecution().GetRunId(), currentMutableState.GetNextEventID(), currentMutableState.IsWorkflowExecutionRunning(), nil
+	runID = currentContext.getExecution().GetRunId()
+	nextEventID = currentMutableState.GetNextEventID()
+	createTaskID = currentMutableState.GetExecutionInfo().CreateTaskID
+	isRunning = currentMutableState.IsWorkflowExecutionRunning()
+	currentRelease(err)
```
Comment (samarabbas) [inline service/history/historyReplicator.go]: You already have a release in a defer function.  Is that really needed?

Comment (wxing1292) [reply service/history/historyReplicator.go]: last time there were panic within the lock, using defer right after getting the lock, will at least make the deadlock go away

**Pull request merged**